File: go/internal/observability/security_event_journal_test.go
package observability

import (
	"os"
	"path/filepath"
	"testing"
)

func TestSecurityEventJournalStartsEmpty(t *testing.T) {
	journal, err := NewSecurityEventJournal(filepath.Join(t.TempDir(), "events.jsonl"))
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	if journal.Size() != 0 {
		t.Fatalf("expected empty journal, got size %d", journal.Size())
	}
}

func TestSecurityEventJournalEmitAndList(t *testing.T) {
	journal, err := NewSecurityEventJournal(filepath.Join(t.TempDir(), "events.jsonl"))
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	ok, reason := journal.Emit(SecurityEvent{
		SchemaVersion: SecurityEventSchemaVersion,
		SourceService: "go-api",
		EventType:     EventSecurityPermissionDenied,
		WorkerID:      "worker-1",
	})
	if !ok {
		t.Fatalf("emit rejected: %s", reason)
	}
	if journal.Size() != 1 {
		t.Fatalf("expected size 1, got %d", journal.Size())
	}
	result := journal.List(SecurityEventListFilters{})
	if len(result.Events) != 1 {
		t.Fatalf("expected 1 listed event, got %d", len(result.Events))
	}
	if result.Events[0].EventID == "" || result.Events[0].Timestamp == "" || result.Events[0].PayloadChecksum == "" {
		t.Fatalf("expected event_id/timestamp/payload_checksum to be assigned: %+v", result.Events[0])
	}
}

func TestSecurityEventJournalRestartPersistence(t *testing.T) {
	path := filepath.Join(t.TempDir(), "events.jsonl")
	journal, err := NewSecurityEventJournal(path)
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	if ok, reason := journal.Emit(SecurityEvent{SchemaVersion: SecurityEventSchemaVersion, SourceService: "go-api", EventType: EventWorkerRegistered}); !ok {
		t.Fatalf("emit rejected: %s", reason)
	}

	restarted, err := NewSecurityEventJournal(path)
	if err != nil {
		t.Fatalf("restart NewSecurityEventJournal: %v", err)
	}
	if restarted.Size() != 1 {
		t.Fatalf("expected restarted journal to reload 1 event, got %d", restarted.Size())
	}
	if restarted.RecoveredLineCount() != 0 {
		t.Fatalf("expected 0 recovered lines on a clean file, got %d", restarted.RecoveredLineCount())
	}
}

func TestSecurityEventJournalCursorPagination(t *testing.T) {
	journal, err := NewSecurityEventJournal(filepath.Join(t.TempDir(), "events.jsonl"))
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	journal.Emit(SecurityEvent{SchemaVersion: SecurityEventSchemaVersion, SourceService: "go-api", EventType: EventWorkerRegistered})
	journal.Emit(SecurityEvent{SchemaVersion: SecurityEventSchemaVersion, SourceService: "go-api", EventType: EventWorkerActivated})

	firstPage := journal.List(SecurityEventListFilters{Limit: 1})
	if len(firstPage.Events) != 1 || firstPage.NextCursor == "" {
		t.Fatalf("expected page of 1 with a non-empty cursor, got %+v", firstPage)
	}
	secondPage := journal.List(SecurityEventListFilters{AfterEventID: firstPage.NextCursor})
	if len(secondPage.Events) != 1 || secondPage.Events[0].EventType != EventWorkerActivated {
		t.Fatalf("expected cursor to resume at the second event, got %+v", secondPage)
	}
}

func TestSecurityEventJournalSeverityFilter(t *testing.T) {
	journal, err := NewSecurityEventJournal(filepath.Join(t.TempDir(), "events.jsonl"))
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	journal.Emit(SecurityEvent{SchemaVersion: SecurityEventSchemaVersion, SourceService: "go-api", EventType: EventWorkerRegistered})
	result := journal.List(SecurityEventListFilters{MinSeverity: SeverityCritical})
	if len(result.Events) != 0 {
		t.Fatalf("expected no events at CRITICAL minimum severity, got %d", len(result.Events))
	}
}

func TestSecurityEventJournalInvalidEventIsDropped(t *testing.T) {
	journal, err := NewSecurityEventJournal(filepath.Join(t.TempDir(), "events.jsonl"))
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	ok, _ := journal.Emit(SecurityEvent{SchemaVersion: SecurityEventSchemaVersion, SourceService: ""})
	if ok {
		t.Fatalf("expected emit of an event with no source_service to be rejected")
	}
	if journal.Size() != 0 {
		t.Fatalf("expected invalid event not to be persisted")
	}
}

func TestSecurityEventJournalCorruptionRecovery(t *testing.T) {
	path := filepath.Join(t.TempDir(), "events.jsonl")
	journal, err := NewSecurityEventJournal(path)
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	journal.Emit(SecurityEvent{SchemaVersion: SecurityEventSchemaVersion, SourceService: "go-api", EventType: EventWorkerRegistered})

	file, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		t.Fatalf("open for corruption append: %v", err)
	}
	file.WriteString("not valid json at all\n")
	file.WriteString(`{"schema_version":1,"event_id":"x"}` + "\n")
	file.Close()

	reloaded, err := NewSecurityEventJournal(path)
	if err != nil {
		t.Fatalf("reload after corruption: %v", err)
	}
	if reloaded.Size() != 1 {
		t.Fatalf("expected 1 valid record to survive corruption, got %d", reloaded.Size())
	}
	if reloaded.RecoveredLineCount() != 2 {
		t.Fatalf("expected 2 recovered lines, got %d", reloaded.RecoveredLineCount())
	}
}

func TestSecurityEventJournalRotationAndRetention(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "events.jsonl")
	journal, err := NewSecurityEventJournal(path)
	if err != nil {
		t.Fatalf("NewSecurityEventJournal: %v", err)
	}
	journal.maxBytes = 200
	journal.maxRetained = 2
	for i := 0; i < 20; i++ {
		journal.Emit(SecurityEvent{
			SchemaVersion: SecurityEventSchemaVersion,
			SourceService: "go-api",
			EventType:     EventHeartbeatAccepted,
			WorkerID:      "worker",
		})
	}
	if _, err := os.Stat(path + ".1"); err != nil {
		t.Fatalf("expected rotation to produce a .1 file: %v", err)
	}
	if _, err := os.Stat(path + ".3"); err == nil {
		t.Fatalf("expected retention count of 2 to never keep a third generation")
	}
	if journal.Size() >= 20 {
		t.Fatalf("expected the active file's in-memory view to no longer hold every event")
	}
}

File: cpp/coordinator/tests/security_event_journal_test.cpp
#include "fl_coordinator/security_event_journal.hpp"
#include "test_support.hpp"

#include <filesystem>
#include <fstream>

namespace fl::coordinator::testing {

void run_security_event_journal_tests(const std::string& scratch_dir) {
    using fl::coordinator::SecurityEvent;
    using fl::coordinator::SecurityEventJournal;
    using fl::coordinator::SecurityEventType;
    using fl::coordinator::SecuritySeverity;
    using fl::coordinator::SecuritySubjectType;

    std::filesystem::remove_all(scratch_dir);
    std::filesystem::create_directories(scratch_dir);
    const std::string journal_path = scratch_dir + "/security_events.jsonl";

    // Emit + round-trip.
    {
        SecurityEventJournal journal(journal_path);
        check(journal.size() == 0, "a fresh journal starts empty");

        SecurityEvent event;
        event.source_service = "coordinator";
        event.source_component = "worker_registry";
        event.event_type = SecurityEventType::kWorkerSuspended;
        event.worker_id = "worker-1";
        event.safe_actor_id = "go-api";
        event.reason_code = "administrative_suspension";
        journal.emit(event);

        check(journal.size() == 1, "emitting one event grows the journal to size 1");

        const auto listed = journal.list({});
        check(listed.events.size() == 1, "list() returns the emitted event");
        check(!listed.events[0].event_id.empty(), "emit() assigns a non-empty event_id");
        check(!listed.events[0].timestamp.empty(), "emit() assigns a non-empty timestamp");
        check(!listed.events[0].payload_checksum.empty(), "emit() assigns a payload_checksum");
        check(listed.events[0].worker_id == "worker-1", "round-tripped event preserves worker_id");
        check(listed.events[0].event_type == SecurityEventType::kWorkerSuspended,
              "round-tripped event preserves event_type");
    }

    // Restart persistence.
    {
        SecurityEventJournal restarted(journal_path);
        check(restarted.size() == 1, "a restarted journal reloads its one persisted event");
        check(restarted.recovered_line_count() == 0,
              "a clean file recovers zero corrupt lines on reload");
    }

    // Cursor pagination.
    {
        SecurityEventJournal journal(journal_path);
        SecurityEvent second;
        second.source_service = "coordinator";
        second.event_type = SecurityEventType::kWorkerActivated;
        second.worker_id = "worker-1";
        journal.emit(second);

        const auto first_page = journal.list({.limit = 1});
        check(first_page.events.size() == 1 && !first_page.next_cursor.empty(),
              "a page smaller than the journal returns a non-empty next_cursor");

        SecurityEventJournal::ListFilters next_filters;
        next_filters.after_event_id = first_page.next_cursor;
        const auto second_page = journal.list(next_filters);
        check(second_page.events.size() == 1 &&
                  second_page.events[0].event_type == SecurityEventType::kWorkerActivated,
              "the cursor from page one correctly resumes at the second event");
    }

    // Filtering.
    {
        SecurityEventJournal journal(journal_path);
        SecurityEventJournal::ListFilters type_filter;
        type_filter.event_type = SecurityEventType::kWorkerActivated;
        const auto filtered = journal.list(type_filter);
        check(filtered.events.size() == 1 &&
                  filtered.events[0].event_type == SecurityEventType::kWorkerActivated,
              "event_type filter returns only matching events");

        SecurityEventJournal::ListFilters severity_filter;
        severity_filter.min_severity = SecuritySeverity::kCritical;
        const auto none = journal.list(severity_filter);
        check(none.events.empty(),
              "a min_severity filter above every emitted event's severity returns nothing");
    }

    // Validation failure is dropped, not thrown or persisted.
    {
        const std::string invalid_path = scratch_dir + "/invalid.jsonl";
        SecurityEventJournal journal(invalid_path);
        SecurityEvent invalid;
        invalid.source_service = "";  // required field missing
        expect_no_throw([&]() { journal.emit(invalid); }, "emitting an invalid event never throws");
        check(journal.size() == 0, "an invalid event is dropped, not persisted");
    }

    // Corruption recovery: a hand-corrupted line is skipped, not fatal.
    {
        const std::string corrupt_path = scratch_dir + "/corrupt.jsonl";
        {
            SecurityEventJournal journal(corrupt_path);
            SecurityEvent event;
            event.source_service = "coordinator";
            event.event_type = SecurityEventType::kWorkerRegistered;
            journal.emit(event);
        }
        {
            std::ofstream file(corrupt_path, std::ios::app | std::ios::binary);
            file << "{not valid json at all\n";
            file << "{\"schema_version\":1,\"event_id\":\"x\"}\n";  // valid JSON, bad
                                                                    // checksum/fields
        }
        SecurityEventJournal reloaded(corrupt_path);
        check(reloaded.size() == 1,
              "a journal with two corrupt trailing lines keeps the one valid record");
        check(reloaded.recovered_line_count() == 2,
              "both corrupt lines are counted as recovered, not fatal");
    }

    // Rotation and retention.
    {
        const std::string rotate_path = scratch_dir + "/rotate.jsonl";
        SecurityEventJournal::Options options;
        options.max_bytes_before_rotation = 200;  // small, forces rotation quickly
        options.max_retained_files = 2;
        SecurityEventJournal journal(rotate_path, options);
        for (int i = 0; i < 20; ++i) {
            SecurityEvent event;
            event.source_service = "coordinator";
            event.event_type = SecurityEventType::kHeartbeatAccepted;
            event.worker_id = "worker-" + std::to_string(i);
            journal.emit(event);
        }
        check(std::filesystem::exists(rotate_path + ".1"), "rotation produces a .1 rotated file");
        check(!std::filesystem::exists(rotate_path + ".3"),
              "retention count of 2 never keeps a third rotated generation");
        check(
            journal.size() < 20,
            "the active file's in-memory view no longer holds every event once rotation occurred");
    }
}

}  // namespace fl::coordinator::testing
